### PR TITLE
Adding r-archive R package

### PR DESCRIPTION
**Pull request by pschloss**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

Used [conda_r_skeleton_helper](https://github.com/bgruening/conda_r_skeleton_helper) to create `meta.yml` and `build.sh` files.

### CODE REVIEW

**Comment by pschloss** [comment]
I wonder if anyone can help me with building this recipe. Initially none of the OS's would build. I added `libarchive` to the `requirements: build:` and `requirements: host:` listings from the output of using `conda_r_skeleton_helper` and that got the linux version to build. Unfortunately, neither os x nor windows are building now. Both of the failed OS's seem to have different error messages. I'm feeling a bit stuck with my skills and googling at this point. If @bgruening or @jdblischak have any suggestions for things to try I'd be very grateful!

### os x
```
installing to /Users/runner/mambaforge/conda-bld/r-archive_1663437129171/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/R/library/00LOCK-work/00new/archive/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
ERROR: loading failed
* removing ‘/Users/runner/mambaforge/conda-bld/r-archive_1663437129171/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/R/library/archive’
Error: package or namespace load failed for ‘archive’:
 .onLoad failed in loadNamespace() for 'archive', details:
  call: dyn.load(lib_path)
  error: unable to load shared object '$SRC_DIR/':
  dlopen($SRC_DIR/, 6): no suitable image found.  Did find:
	$SRC_DIR/: not a file
	$SRC_DIR: not a file
Error: loading failed
Execution halted
Traceback (most recent call last):
  File "/Users/runner/work/1/s/.ci_support/build_all.py", line 223, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "/Users/runner/work/1/s/.ci_support/build_all.py", line 120, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "/Users/runner/work/1/s/.ci_support/build_all.py", line 179, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/Users/runner/mambaforge/lib/python3.10/site-packages/conda_build/api.py", line 186, in build
    return build_tree(
  File "/Users/runner/mambaforge/lib/python3.10/site-packages/conda_build/build.py", line 3091, in build_tree
    packages_from_this = build(metadata, stats,
  File "/Users/runner/mambaforge/lib/python3.10/site-packages/conda_build/build.py", line 2214, in build
    utils.check_call_env(cmd, env=env, rewrite_stdout_env=rewrite_env,
  File "/Users/runner/mambaforge/lib/python3.10/site-packages/conda_build/utils.py", line 412, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/Users/runner/mambaforge/lib/python3.10/site-packages/conda_build/utils.py", line 392, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/Users/runner/mambaforge/conda-bld/r-archive_1663437129171/work/conda_build.sh']' returned non-zero exit status 1.
```

### windows
```
g++ -shared -s -o archive.dll tmp.def archive.o archive_extract.o archive_read.o archive_write.o archive_write_direct.o archive_write_files.o cpp11.o r_archive.o -L../windows/libarchive-3.6.1/lib-5.3.0/x64 -L../windows/libarchive-3.6.1/lib/x64 -larchive -lcrypto -lnettle -lregex -lexpat -llzo2 -llzma -llz4 -lbz2 -lzstd -lz -liconv -lbcrypt -LC:/bld/r-archive_1663437177396/_h_env/lib/R/bin/x64 -lR
../windows/libarchive-3.6.1/lib/x64/libarchive.a(archive_write_set_format_iso9660.o):(.text$set_date_time+0x1e5): undefined reference to `__imp__get_timezone'
../windows/libarchive-3.6.1/lib/x64/libarchive.a(archive_write_set_format_iso9660.o):(.text$set_time_915+0x54): undefined reference to `__imp__get_timezone'
../windows/libarchive-3.6.1/lib/x64/libarchive.a(archive_read_disk_windows.o):(.text$tree_reopen+0x28d): undefined reference to `__imp___acrt_iob_func'
../windows/libarchive-3.6.1/lib/x64/libarchive.a(archive_read_disk_windows.o):(.text$tree_reopen+0x2e5): undefined reference to `__imp___acrt_iob_func'
../windows/libarchive-3.6.1/lib/x64/libarchive.a(archive_write_disk_windows.o):(.text$la_GetFunctionKernel32+0x5a): undefined reference to `__imp___acrt_iob_func'
../windows/libarchive-3.6.1/lib/x64/libexpat.a(xmlparse.o):(.text+0x1800): undefined reference to `__imp___acrt_iob_func'
../windows/libarchive-3.6.1/lib/x64/libbz2.a(bzlib.o):(.text+0x689): undefined reference to `__imp___acrt_iob_func'
../windows/libarchive-3.6.1/lib/x64/libbz2.a(bzlib.o):(.text+0x111a): more undefined references to `__imp___acrt_iob_func' follow
collect2.exe: error: ld returned 1 exit status
gcc  -I"C:/bld/r-archive_1663437177396/_h_env/lib/R/include" -DNDEBUG -I../windows/libarchive-3.6.1/include -I. -I'C:/bld/r-archive_1663437177396/_h_env/lib/R/library/cli/include' -I'C:/bld/r-archive_1663437177396/_h_env/lib/R/library/cpp11/include'        -O2 -Wall  -std=gnu99 -march=x86-64 -mtune=generic -c connection/connection.c -o connection/connection.o
mkdir -p ../inst/lib/x64/
g++  -Wl,--export-all-symbols -shared  -L"C:/bld/r-archive_1663437177396/_h_env/lib/R/bin/x64" -lR -o ../inst/lib/x64/libconnection.dll connection/connection.o
connection/connection.o:connection.c:(.text+0x1e): undefined reference to `R_GetConnection'
connection/connection.o:connection.c:(.text+0x1): undefined reference to `R_new_custom_connection'
connection/connection.o:connection.c:(.text+0x33): undefined reference to `R_ReadConnection'
collect2.exe: error: ld returned 1 exit status
make: *** [../inst/lib/x64/libconnection.dll] Error 1
Makevars.win:20: recipe for target '../inst/lib/x64/libconnection.dll' failed
ERROR: compilation failed for package 'archive'
* removing 'C:/bld/r-archive_1663437177396/_h_env/lib/R/library/archive'

(%BUILD_PREFIX%) %SRC_DIR%>IF 1 NEQ 0 exit /B 1 
Traceback (most recent call last):
  File "D:\a\1\s\.ci_support\build_all.py", line 223, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "D:\a\1\s\.ci_support\build_all.py", line 120, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "D:\a\1\s\.ci_support\build_all.py", line 179, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "C:\Miniforge\lib\site-packages\conda_build\api.py", line 186, in build
    return build_tree(
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 3091, in build_tree
    packages_from_this = build(metadata, stats,
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 2182, in build
    windows.build(m, build_file, stats=build_stats, provision_only=provision_only)
  File "C:\Miniforge\lib\site-packages\conda_build\windows.py", line 297, in build
    check_call_env(cmd, cwd=m.config.work_dir, stats=stats, rewrite_stdout_env=rewrite_env)
  File "C:\Miniforge\lib\site-packages\conda_build\utils.py", line 412, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "C:\Miniforge\lib\site-packages\conda_build\utils.py", line 392, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['cmd.exe', '/d', '/c', 'conda_build.bat']' returned non-zero exit status 1.
##[error]Cmd.exe exited with code '1'.
```

**Comment by jdblischak** [comment]
It's not a noarch recipe. It has C++ code to compile in [`src`](https://github.com/r-lib/archive/tree/main/src). That's why the conda skeleton included compilers in the build dependencies.

**Comment by pschloss** [comment]
It would be great if @conda-forge/help-r could check out this PR for r-archive. It passes although we did skip building windows version.

File: recipes/r-archive/0001-Force-so-extension-on-libconnection.patch [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+--- a/R/zzz.R
++++ b/R/zzz.R
+@@ -1,5 +1,11 @@
+ .onLoad <- function(libname, pkgname) {
+   lib_path <- system.file("lib", .Platform$r_arch, paste0("libconnection", .Platform$dynlib.ext), package = "archive")
++
++  if(lib_path == ""){
++    lib_path <- system.file("lib", .Platform$r_arch, paste0("libconnection", ".so"), package = 
++"archive")
++  }
```
Comment (jdblischak) [inline recipes/r-archive/0001-Force-so-extension-on-libconnection.patch:21]: Could you please provide some more context on why you created this patch? Also, below in this file the function `.onUnload()` also uses `.Platform$dynlib.ext`, so presumably it should be changed there as well

Comment (pschloss) [reply recipes/r-archive/0001-Force-so-extension-on-libconnection.patch]: When I am building this using the default Mac OS X environment, `.Platform$dynlib.ext` is `.so` and so it happily finds `libconnection.so` and generates a full string for `lib_path`.

```
lib_path <- system.file("lib", .Platform$r_arch, paste0("libconnection", .Platform$dynlib.ext), package = "archive")
```

However, when I build this in a conda environment, `.Platform$dynlib.ext` is `.dylib`. This leads `lib_path` to being `""`, not finding `libconnection.so` and failing to load the library.

The solution that I've come up with is to replace those lines with the co-submitted patch. I've tested this on a Mac OS X with R (>4.1) inside and outside of conda and it works. I have updated the `.onUnload()` function as well.

Comment (jdblischak) [reply recipes/r-archive/0001-Force-so-extension-on-libconnection.patch]: Thanks for the explanation. I don't understand why `.Platform$dynlib.ext` is different inside a conda env on macOS. Hopefully someone more knowledgeable will be able to comment on the suitability of your patch. It's hard for me to anticipate any unintended consequences.

**Pull request closed without merging**